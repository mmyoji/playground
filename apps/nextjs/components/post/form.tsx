import { type FormEventHandler, useState } from "react";
import { toast } from "react-toastify";

export type IOnSubmit = (
  args: {
    title: string;
    published: boolean;
  },
) => Promise<void>;

interface Props {
  buttonText: string;
  init?: {
    title: string;
    published: boolean;
  };
  onSubmit: IOnSubmit;
}

interface FormProps {
  title: string;
  published: boolean;
}

export function PostForm({ buttonText, init, onSubmit }: Props) {
  const [formData, setFormData] = useState<FormProps>({
    title: init?.title ?? "",
    published: init?.published ?? false,
  });

  const handleSubmit: FormEventHandler<HTMLFormElement> = async (event) => {
    event.preventDefault();

    const { title } = formData;

    if (!title.trim()) {
      toast.error("`Title` must be present!");
      return;
    }

    await onSubmit(formData);
  };

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <label htmlFor="title">Title</label>
        <br />
        <input
          name="title"
          id="title"
          data-testid="title"
          value={formData.title}
          required
          onChange={({ target: { value } }) => {
            setFormData((prev) => ({ ...prev, title: value }));
          }}
        />
      </div>

      <div>
        <label htmlFor="published">
          <input
            type="checkbox"
            name="published"
            id="published"
            data-testid="published"
            checked={formData.published}
            onChange={() => {
              setFormData((prev) => ({ ...prev, published: !prev.published }));
            }}
          />
          Published
        </label>
      </div>

      <input type="submit" data-testid="submit" value={buttonText} />
    </form>
  );
}
