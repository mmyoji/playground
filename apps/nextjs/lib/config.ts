const PORT = Number(process.env.NEXT_PUBLIC_PORT || "3000");

export const config = {
  app: {
    name: `Blog App`,
  },
  port: PORT,
  host: `http://localhost:${PORT}`,
};
