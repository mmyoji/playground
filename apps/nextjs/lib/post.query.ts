import { Prisma } from "@prisma/client";
import { prisma } from "@/prisma/client";

async function fetchMany(
  { after, take = 10 }: { take?: number; after?: string },
) {
  const select: Prisma.PostSelect = {
    id: true,
    title: true,
    updatedAt: true,
  };
  const where: Prisma.PostWhereInput = {
    published: true,
  };
  const orderBy: Prisma.PostOrderByWithRelationInput = {
    updatedAt: "desc",
  };

  const args: Prisma.PostFindManyArgs = after
    ? {
      select,
      where,
      orderBy,
      take,
      skip: 1,
      cursor: {
        id: after,
      },
    }
    : {
      select,
      where,
      orderBy,
      take,
    };

  const posts = await prisma.post.findMany(args);
  return posts.map((post) => ({
    ...post,
    updatedAt: post.updatedAt.toLocaleString(),
  }));
}

async function fetchOne(
  { id, onlyPublished = true }: { id: string; onlyPublished?: boolean },
) {
  const where: Prisma.PostWhereInput = onlyPublished
    ? {
      id,
      published: true,
    }
    : {
      id,
    };

  const post = await prisma.post.findFirst({
    where,
  });
  if (!post) {
    return null;
  }

  return ({
    ...post,
    createdAt: post.createdAt.toLocaleString(),
    updatedAt: post.updatedAt.toLocaleString(),
  });
}

export const postQuery = {
  fetchMany,
  fetchOne,
};
