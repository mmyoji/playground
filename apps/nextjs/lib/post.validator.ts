export type Params = {
  title: string;
  published: boolean;
};

async function validate(params: Partial<Params>): Promise<boolean> {
  if (!("title" in params) || !("published" in params)) {
    return false;
  }

  const { title, published } = params;

  return typeof title === "string" && !!title && typeof published === "boolean";
}

export const postValidator = {
  validate,
};
