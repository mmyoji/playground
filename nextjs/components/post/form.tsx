import { type FormEventHandler, useState } from "react";
import { toast } from "react-toastify";

import { Checkbox } from "@/components/shared/checkbox";
import { Label } from "@/components/shared/label";
import { SubmitButton } from "@/components/shared/submit-button";
import { TextInput } from "@/components/shared/text-input";

export type IOnSubmit = (
  args: {
    title: string;
    published: boolean;
  },
) => Promise<void>;

interface Props {
  buttonText: string;
  init?: {
    title: string;
    published: boolean;
  };
  onSubmit: IOnSubmit;
}

interface FormProps {
  title: string;
  published: boolean;
}

export function PostForm({ buttonText, init, onSubmit }: Props) {
  const [formData, setFormData] = useState<FormProps>({
    title: init?.title ?? "",
    published: init?.published ?? false,
  });

  const handleSubmit: FormEventHandler<HTMLFormElement> = async (event) => {
    event.preventDefault();

    const { title } = formData;

    if (!title.trim()) {
      toast.error("`Title` must be present!");
      return;
    }

    await onSubmit(formData);
  };

  return (
    <form onSubmit={handleSubmit} className="my-2">
      <div>
        <Label htmlFor="title">Title</Label>
        <br />
        <TextInput
          id="title"
          value={formData.title}
          required
          onChange={(value) => {
            setFormData((prev) => ({ ...prev, title: value }));
          }}
        />
      </div>

      <div>
        <Checkbox
          id="published"
          checked={formData.published}
          label="Published"
          onChange={() => {
            setFormData((prev) => ({ ...prev, published: !prev.published }));
          }}
        />
      </div>

      <div className="my-2">
        <SubmitButton label={buttonText} />
      </div>
    </form>
  );
}
