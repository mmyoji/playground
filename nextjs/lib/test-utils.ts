import userEvent from "@testing-library/user-event";

import { cleanup, render } from "@testing-library/react";
import { ReactElement } from "react";

function setup(jsx: ReactElement) {
  return {
    user: userEvent.setup(),
    ...render(jsx),
  };
}

function teardown() {
  cleanup();
}

export { setup, teardown };
