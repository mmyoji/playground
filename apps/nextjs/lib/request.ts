import { config } from "./config";

const RequestFailedError = new Error("request failed.");

type MutateMethod = "POST" | "PUT" | "PATCH" | "DELETE";

export async function command<T extends Record<string, unknown>>(
  path: string,
  method: MutateMethod,
  data?: T,
): Promise<{ error?: Error }> {
  const init: RequestInit = {
    method,
    headers: {
      "Content-Type": "application/json",
    },
  };

  if (data) {
    init.body = JSON.stringify(data);
  }

  const res = await fetch(`${config.host}${path}`, init);
  if (res.ok) {
    return {};
  }

  return {
    error: RequestFailedError,
  };
}
