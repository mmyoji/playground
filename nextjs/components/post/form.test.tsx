import { afterEach, describe, expect, it, vi } from "vitest";
import { waitFor } from "@testing-library/react";
import { ComponentProps } from "react";
import { toast } from "react-toastify";
import { setup, teardown } from "@/lib/test-utils";
import { IOnSubmit, PostForm } from "./form";

vi.mock("react-toastify", () => {
  return {
    toast: {
      error: vi.fn(),
    },
  };
});

const dummyHandleSubmit: IOnSubmit = (_args) => Promise.resolve(undefined);

describe("<PostForm>", () => {
  const _setup = (props: Partial<ComponentProps<typeof PostForm>>) => {
    const mockHandleSubmit = props.onSubmit ||
      vi.fn().mockImplementation(dummyHandleSubmit);
    return {
      mockHandleSubmit,
      ...setup(
        <PostForm
          {...props}
          buttonText={props.buttonText || "Submit"}
          onSubmit={mockHandleSubmit}
        />,
      ),
    };
  };

  afterEach(() => {
    teardown();
    vi.clearAllMocks();
  });

  it("matches former snapshot", async () => {
    const { container } = _setup({});

    expect(container).toMatchSnapshot();
  });

  it("renders a form", () => {
    const buttonText = "Click Me!";
    const { getByLabelText, getByDisplayValue, getByTestId } = _setup({
      buttonText,
    });

    const titleInput = getByTestId("title");
    const publishedInput = getByTestId("published");

    expect(getByLabelText("Title")).toBeDefined();
    expect(getByLabelText("Published")).toBeDefined();
    expect(titleInput.getAttribute("value")).toBe("");
    expect(publishedInput.hasAttribute("checked")).toBe(false);
    expect(getByDisplayValue(buttonText)).toBeDefined();
  });

  it("emits error message when values are invalid", async () => {
    const { mockHandleSubmit, user, getByTestId } = _setup({});

    const titleInput = getByTestId("title");
    await user.type(titleInput, " ");

    const submitButton = getByTestId("submit");
    await user.click(submitButton);

    await waitFor(() => {
      expect(toast.error).toBeCalledTimes(1);
      expect(toast.error).toBeCalledWith("`Title` must be present!");
    });
    expect(mockHandleSubmit).toBeCalledTimes(0);
  });

  it("submits form data", async () => {
    const { mockHandleSubmit, user, getByTestId } = _setup({});

    const titleInput = getByTestId("title");
    await user.type(titleInput, "My Title");

    const publishedInput = getByTestId("published");
    await user.click(publishedInput);

    const submitButton = getByTestId("submit");
    await user.click(submitButton);

    await waitFor(() => {
      expect(mockHandleSubmit).toBeCalledTimes(1);
      expect(mockHandleSubmit).toBeCalledWith({
        title: "My Title",
        published: true,
      });
    });
    expect(toast.error).toBeCalledTimes(0);
  });

  it("sets initial values when init is passed", async () => {
    const init = {
      title: "Dr. Stone",
      published: true,
    };
    const { getByTestId } = _setup({ init });

    const titleInput = getByTestId("title");
    const publishedInput = getByTestId("published");

    expect(titleInput.getAttribute("value")).toBe(init.title);
    expect(publishedInput.hasAttribute("checked")).toBe(true);
  });
});
