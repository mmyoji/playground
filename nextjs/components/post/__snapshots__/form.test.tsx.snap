// Vitest Snapshot v1

exports[`<PostForm> > matches former snapshot 1`] = `
<div>
  <form
    class="my-2"
  >
    <div>
      <label
        class="font-bold"
        for="title"
      >
        Title
      </label>
      <br />
      <input
        class="px-3 py-1.5 border border-solid border-gray-300 rounded"
        data-testid="title"
        id="title"
        name="title"
        required=""
        value=""
      />
    </div>
    <div>
      <label
        for="published"
      >
        <input
          data-testid="published"
          id="published"
          name="published"
          type="checkbox"
        />
        <span
          class="ml-1"
        >
          Published
        </span>
      </label>
    </div>
    <div
      class="my-2"
    >
      <input
        class="p-2 text-green-900 hover:text-green-700 bg-emerald-200 hover:bg-emerald-100 rounded-md border-2 border-solid border-emerald-200 hover:border-emerald-100"
        data-testid="submit"
        type="submit"
        value="Submit"
      />
    </div>
  </form>
</div>
`;
